include/lsst/afw/table/io/FitsReader.h

```diff
@@ -15,135 +17,85 @@ namespace lsst { namespace afw { namespace table { namespace io {
+        ContainerT container(boost::dynamic_pointer_cast<typename ContainerT::Table>(table));
+        if (!container.getTable()) {
+            throw LSST_EXCEPT(
+                pex::exceptions::RuntimeError,
+                "Invalid table class for catalog."
+            );
+        }
+        std::size_t nRows = fits.countRows();
+        container.reserve(nRows);
+        for (std::size_t row = 0; row < nRows; ++row) {
+            mapper.readRecord(
+                // We need to be able to support reading Catalog<T const>, since it shares the same template
+                // as Catalog<T> (which invokes this method in readFits).
+                const_cast<typename boost::remove_const<typename ContainerT::Record>::type&>(
+                    *container.addNew()
+                ),
+                fits, row
+            );
+        }
+        return container;
     }
 
-    /// @brief Low-level entry point for reading FITS files into arbitrary containers.
-    template <typename ContainerT>
-    static ContainerT apply(Fits & fits, int flags=0) {
-        PTR(FitsReader) reader = make(&fits, PTR(io::InputArchive)(), flags);
-        return reader->template read<ContainerT>();
+    template <typename ContainerT, typename SourceT>
+    static ContainerT apply(SourceT & source, int hdu, int ioFlags, PTR(InputArchive) archive=nullptr) {
+        afw::fits::Fits fits(source, "r", afw::fits::Fits::AUTO_CLOSE | afw::fits::Fits::AUTO_CHECK);
+        fits.setHdu(hdu);
+        return apply<ContainerT>(fits, ioFlags, archive);
     }
 
-    /**
-     *  @brief Construct from a wrapped cfitsio pointer and (ignored) InputArchive.
-     *
-     *  Subclasses that require an InputArchive should accept the one that is passed in,
-     *  but may need to construct their own from the HDUs following the catalog HDU(s)
-     *  if this pointer is null.
-     */
-    explicit FitsReader(Fits * fits, PTR(InputArchive), int flags) : _fits(fits), _flags(flags) {}
-
-protected:
-
-    /// @copydoc Reader::_readTable
-    virtual PTR(BaseTable) _readTable();
-
-    /// @copydoc Reader::_readRecord
-    virtual PTR(BaseRecord) _readRecord(PTR(BaseTable) const & table);
+    virtual PTR(BaseTable) makeTable(
+        FitsSchemaInputMapper & mapper,
+        PTR(daf::base::PropertyList) metadata,
+        int ioFlags,
+        bool stripMetadata
+    ) const;
 
-    /// @brief Should be called by any reimplementation of _readTable.
-    void _startRecords(BaseTable & table);
+    virtual bool usesArchive(int ioFlags) const { return false; }
 
-    struct ProcessRecords;
+    virtual ~FitsReader() {}
 
-    Fits * _fits;         // cfitsio pointer in a conveniencer wrapper
-    int _flags;           // subclass-defined flags to control FITS reading
-    std::size_t _row;     // which row we're currently reading
 private:
 
-    friend class afw::table::Schema;
+    friend class Schemaa;
```

Comment:
Interesting.  Looks like that friend declaration isn't needed anymore, and I forgot to remove it.